### DISCUSSION TITLE
NNAPI flag ignored

### DISCUSSION THREAD

**Discussion by marcinkeviciusp** [microsoft/onnxruntime#9177] [Other Q&A] [answered]
I've been trying to benchmark NNAPI as compared to CPU and some other frameworks, but whenever I try to make it use NNAPI (`options.addNnapi(flags)` in Java or `OrtSessionOptionsAppendExecutionProvider_Nnapi(session_options, nnapi_flags)` in C++) the performance stays identical to CPU - in fact, changing number of of threads or any other parameters (e.g. batch size, quantization) affects the resulting time identically as well - this leads me to believe that it's still just using / falling back on CPU.

In the Java project I used `implementation 'com.microsoft.onnxruntime:onnxruntime-mobile:1.8.2'`, while for C++ I built the library (with --use_nnapi) myself.

Is there any way I can debug this (I've not found any ways to output which EP is actually used)? Or perhaps there's some known issue I'm not aware of? It seems like some combination of all these attempts should have worked...

I have tested with:
- Model: ResNet50 (quantized and unquantized; .onnx and .ort (converted with and without --use_nnapi flag - not using the flag does make it crash on inference telling to use the flag))
- Device: Samsung S20 (which, according to my research, has an NPU, which NNAPI should utilize).

Edit: I should also mention I've tried all NNAPI flag coombinations (enable fp16, disable cpu, nchw) which had no effect.

**Comment by edgchen1** [comment, marked as answer]
There are some log messages that may be useful. You may need to set the log level to a more verbose level to see them.

Statistics about how many nodes the NNAPI EP supports (INFO log level):
https://github.com/microsoft/onnxruntime/blob/430e80e7b6e5e6222b2d90ca5e43609d62082722/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L200-L212

Information about node EP assignment (VERBOSE log level):
https://github.com/microsoft/onnxruntime/blob/430e80e7b6e5e6222b2d90ca5e43609d62082722/onnxruntime/core/session/inference_session.cc#L908-L920

**Comment by guoyu-wang** [comment]
ResNet50 is support by our NNAPI execution provider, and can take advantage of the hardware accelerators in Samsung s20. 
However, NNAPI execution provider does not support models with dynamic input shapes, this is probably the reason that none of the NNAPI options worked since the execution always falls back to CPU execution provider.
Please try the same ResNet50 model with static input shape.

**Comment by vpenades** [comment]
I am having similar behavior, my model's performance is the same regardless of the flags I use; input is fixed size and NCHW

It could be good to have a general "tips & tricks" to optimize performance on Android/NNAPI
